util/collections/src/garbage/atomic/counted.rs

```diff
@@ -61,15 +61,36 @@ impl GcCounter {
 		let slot = slot_timestamp % self.num_slots;
 		let (active_slot_timestamp, count) = &self.value_lifetimes[slot as usize];
 
-		// Atomically check and set the timestamp if it doesn't match current_time
+		// Atomically check and set the timestamp if it doesn't match the current slot
 		let active_slot = active_slot_timestamp.load(Ordering::Relaxed);
+		let active_amount = count.load(Ordering::Relaxed);
+
 		if active_slot == slot {
-			// Same timestamp, increment count
+			// Same timestamp, increment count.
+			// At worst, this adds to a later slot.
+			// But, such should not happen under safe usage of this API.
+			// Unless gc windows are very small.
 			count.fetch_add(amount, Ordering::SeqCst);
 		} else {
-			// Different timestamp, reset slot for new period
-			active_slot_timestamp.store(slot_timestamp, Ordering::SeqCst);
-			count.store(amount, Ordering::SeqCst);
+			// now we will zero out the slot and add the new value
+			// Use compare_exchange to safely reset the slot and count only
+			if active_slot_timestamp
+				.compare_exchange(active_slot, slot_timestamp, Ordering::SeqCst, Ordering::Relaxed)
```

Comment:
```rust
compare_exchange(active_slot, slot_timestamp, Ordering::SeqCst, Ordering::Relaxed)
```

We want total ordering on the successful operation as we are under the hood performing a store operation that can clobber the other values. We don't want to allow for any disagreement about the order of these operations. 

The relaxed ordering on the failure case is because we actual don't use the failure case [load](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html). If the comparison fails, we just decide we're going to add to the value. So, we don't need to impose any synchronization here.